From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: satvrn <pastawho@gmail.com>
Date: Tue, 26 Sep 2023 20:09:46 -0700
Subject: [PATCH] Krypton: network.microopt


diff --git a/src/main/java/net/minecraft/network/Utf8String.java b/src/main/java/net/minecraft/network/Utf8String.java
index 7b39060cde0ec31c4adf36960b33cefcd5138527..783cd68b7db5a952085068561ff22c4ab84ccf12 100644
--- a/src/main/java/net/minecraft/network/Utf8String.java
+++ b/src/main/java/net/minecraft/network/Utf8String.java
@@ -31,6 +31,8 @@ public class Utf8String {
     }
 
     public static void write(ByteBuf buf, CharSequence string, int length) {
+        // Krypton start - optimized version
+        /*
         if (string.length() > length) {
             throw new EncoderException("String too big (was " + string.length() + " characters, max " + length + ")");
         } else {
@@ -51,5 +53,22 @@ public class Utf8String {
             }
 
         }
+        */
+        // Krypton end
+
+        // Krypton start - network.microopt
+        // Mojang _almost_ gets it right, but stumbles at the finish line...
+        if (string.length() > length) {
+            throw new EncoderException("String too big (was " + string.length() + " characters, max " + length + ")");
+        }
+        int utf8Bytes = ByteBufUtil.utf8Bytes(string);
+        int maxBytesPermitted = ByteBufUtil.utf8MaxBytes(length);
+        if (utf8Bytes > maxBytesPermitted) {
+            throw new EncoderException("String too big (was " + utf8Bytes + " bytes encoded, max " + maxBytesPermitted + ")");
+        } else {
+            VarInt.write(buf, utf8Bytes);
+            buf.writeCharSequence(string, StandardCharsets.UTF_8);
+        }
+        // Krypton end
     }
 }
