From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: satvrn <pastawho@gmail.com>
Date: Wed, 27 Sep 2023 19:38:55 -0700
Subject: [PATCH] Krypton: change Netty PooledByteBufAllocator arena size


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index ec268189b19b6fa5c4521f96ce211a531db35ec5..751ef8c1738c87e533715f279565b32deee2a575 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -73,6 +73,21 @@ import org.bukkit.configuration.file.YamlConfiguration;
 
 public class Main {
 
+    // Krypton start - change Netty PooledByteBufAllocator arena size
+    static {
+        // By default, Netty allocates 16MiB arenas for the PooledByteBufAllocator. This is too much
+        // memory for Minecraft, which imposes a maximum packet size of 2MiB! We'll use 4MiB as a more
+        // sane default.
+        //
+        // Note: io.netty.allocator.pageSize << io.netty.allocator.maxOrder is the formula used to
+        // compute the chunk size. We lower maxOrder from its default of 11 to 9. (We also use a null
+        // check, so that the user is free to choose another setting if need be.)
+        if (System.getProperty("io.netty.allocator.maxOrder") == null) {
+            System.setProperty("io.netty.allocator.maxOrder", "9");
+        }
+    }
+    // Krypton end
+
     private static final Logger LOGGER = LogUtils.getLogger();
 
     public Main() {}
