From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mariell Hoversholm <proximyst@proximyst.com>
Date: Thu, 30 Jul 2020 11:57:24 +0200
Subject: [PATCH] Projectiles die after 60 sec and in portals

---
 .../java/net/minecraft/server/IProjectile.java     | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/main/java/net/minecraft/server/IProjectile.java b/src/main/java/net/minecraft/server/IProjectile.java
index 9f5ce64a60fe7c312399ee416b11b84213dd3bee..f860647d5ba848330aae9d75dddbc17b2e15ef99 100644
--- a/src/main/java/net/minecraft/server/IProjectile.java
+++ b/src/main/java/net/minecraft/server/IProjectile.java
@@ -67,6 +67,11 @@ public abstract class IProjectile extends Entity {
 
     @Override
     public void tick() {
+        // Papercut start - despawn projectiles after 60 seconds
+        if (ticksLived >= 60 * 20) {
+            die();
+        }
+        // Papercut end
         if (!this.d) {
             this.d = this.h();
         }
@@ -74,6 +79,15 @@ public abstract class IProjectile extends Entity {
         super.tick();
     }
 
+    // Papercut start - projectiles going through portals get despawned
+    @Override
+    protected void doPortalTick() {
+        if (this.inPortal && !(this instanceof EntityThrownTrident)) {
+            die();
+        } else if (isAlive() && this.valid) super.doPortalTick();
+    }
+    // Papercut end
+
     private boolean h() {
         Entity entity = this.getShooter();
 
