From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Fri, 3 Sep 2021 18:41:27 +0100
Subject: [PATCH] Crunchy Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 83d61dfe..fd8738bb 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1737,6 +1737,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Graphene end

+        // Crunchy start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCrunchyConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Crunchy end
+
         /**
          * Sends the component to the player
          *
