From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Mon, 13 Dec 2021 11:28:08 -0500
Subject: [PATCH] skip event if no listeners


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 780b3777272789207042d0a6ff18bd88311672b0..3da8d521d0d10d9a997b652c6c32c0b2a6f66265 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -656,6 +656,14 @@ public final class SimplePluginManager implements PluginManager {
      */
     @Override
     public void callEvent(@NotNull Event event) {
+        // JettPack start
+        HandlerList handlers = event.getHandlers();
+        RegisteredListener[] listeners = handlers.getRegisteredListeners();
+        if (listeners.length == 0) {
+            return;
+        }
+        // JettPack end
+
         // Paper - replace callEvent by merging to below method
         // KTP start - optimize spigot event bus
         final var isAsync = event.isAsynchronous();
@@ -669,8 +677,8 @@ public final class SimplePluginManager implements PluginManager {
         }
         // KTP end - optimize spigot event bus

-        HandlerList handlers = event.getHandlers();
-        RegisteredListener[] listeners = handlers.getRegisteredListeners();
+        // HandlerList handlers = event.getHandlers();
+        // RegisteredListener[] listeners = handlers.getRegisteredListeners();

         for (RegisteredListener registration : listeners) {
             if (!registration.getPlugin().isEnabled()) {

