From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Wed, 20 Oct 2021 22:30:55 +0100
Subject: [PATCH] Server Reload Events


diff --git a/src/main/java/dev/mrfishcakes/crunchy/event/server/PostServerReloadEvent.java b/src/main/java/dev/mrfishcakes/crunchy/event/server/PostServerReloadEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..025aa2c655af8acfba91b3e1f699e7b1a0f2d8ab
--- /dev/null
+++ b/src/main/java/dev/mrfishcakes/crunchy/event/server/PostServerReloadEvent.java
@@ -0,0 +1,23 @@
+package dev.mrfishcakes.crunchy.event.server;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Event that is called when the server reload is complete
+ */
+public class PostServerReloadEvent extends Event {
+
+    private static final HandlerList HANDLERS = new HandlerList();
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLERS;
+    }
+
+    public static HandlerList getHandlerList() {
+        return HANDLERS;
+    }
+
+}
diff --git a/src/main/java/dev/mrfishcakes/crunchy/event/server/ServerReloadEvent.java b/src/main/java/dev/mrfishcakes/crunchy/event/server/ServerReloadEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..74ae086fdc4ebd86f8b3fdcf0e98b672235427a7
--- /dev/null
+++ b/src/main/java/dev/mrfishcakes/crunchy/event/server/ServerReloadEvent.java
@@ -0,0 +1,23 @@
+package dev.mrfishcakes.crunchy.event.server;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Event that is called when the server reloads
+ */
+public class ServerReloadEvent extends Event {
+
+    private static final HandlerList HANDLERS = new HandlerList();
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return HANDLERS;
+    }
+
+    public static HandlerList getHandlerList() {
+        return HANDLERS;
+    }
+
+}
