From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Mon, 13 Dec 2021 11:28:08 -0500
Subject: [PATCH] skip event if no listeners


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 5463b8be..f4513214 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -613,6 +613,14 @@ public final class SimplePluginManager implements PluginManager {
      */
     @Override
     public void callEvent(@NotNull Event event) {
+        // JettPack start
+        HandlerList handlers = event.getHandlers();
+        RegisteredListener[] listeners = handlers.getRegisteredListeners();
+        if (listeners.length == 0) {
+            return;
+        }
+        // JettPack end
+
         // Paper - replace callEvent by merging to below method
         // KTP start - optimize spigot event bus
         final boolean isAsync = event.isAsynchronous();
@@ -624,8 +632,8 @@ public final class SimplePluginManager implements PluginManager {
         }
         // KTP end - optimize spigot event bus

-        HandlerList handlers = event.getHandlers();
-        RegisteredListener[] listeners = handlers.getRegisteredListeners();
+        //HandlerList handlers = event.getHandlers();
+        //RegisteredListener[] listeners = handlers.getRegisteredListeners();

         for (RegisteredListener registration : listeners) {
             if (!registration.getPlugin().isEnabled()) {
