From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Satyrn <satyrn@disroot.org>
Date: Sun, 5 Dec 2021 21:24:00 +0000
Subject: [PATCH] Bump Netty and Velocity


diff --git a/build.gradle.kts b/build.gradle.kts
index f9b1892d9..de0cadf01 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -52,10 +52,10 @@ dependencies {
     runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.0")
     runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.0")

-    implementation("io.netty:netty-all:4.1.65.Final") // Paper
+    implementation("io.netty:netty-all:4.1.70.Final") // Interstellar

     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
-    implementation("com.velocitypowered:velocity-native:1.1.0-SNAPSHOT") // Paper
+    implementation("com.velocitypowered:velocity-native:1.1.9") // Interstellar

     implementation("com.github.technove:AIR:fe3dbb4420") // Airplane - config
     implementation("org.yaml:snakeyaml:1.28")
