From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 10 Nov 2021 20:18:43 -0500
Subject: [PATCH] configurable fishing rod length


diff --git a/src/main/java/me/titaniumtown/JettPackConfig.java b/src/main/java/me/titaniumtown/JettPackConfig.java
index 2217bc4a378866eaf224731685b759a27268b655..56e9e776750ad569de0d69a3f96994fc4e64d555 100644
--- a/src/main/java/me/titaniumtown/JettPackConfig.java
+++ b/src/main/java/me/titaniumtown/JettPackConfig.java
@@ -116,7 +116,9 @@ public class JettPackConfig {
     }
 
     public static boolean logPlayerLoginLoc = true;
+    public static double fishingRodLength = 1024.0D;
     private static void general() {
         logPlayerLoginLoc = getBoolean("settings.log-player-login-location", logPlayerLoginLoc);
+        fishingRodLength = getDouble("settings.fishing-rod-length", fishingRodLength);
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 1037d0a0cdd4fd7aa99a958ee969759c5883fdc0..89213e3b8ea52d1aec8daee04f26a2a0e4636f6c 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -143,7 +143,7 @@ public class FishingHook extends Projectile {
 
     @Override
     public boolean shouldRenderAtSqrDistance(double distance) {
-        double d1 = 64.0D;
+        // double d1 = 64.0D; // Jettpack
 
         return distance < 4096.0D;
     }
@@ -248,7 +248,7 @@ public class FishingHook extends Projectile {
                 this.setDeltaMovement(Vec3.ZERO);
             }
 
-            double d1 = 0.92D;
+            // double d1 = 0.92D; // Jettpack
 
             this.setDeltaMovement(this.getDeltaMovement().scale(0.92D));
             this.reapplyPosition();
@@ -266,7 +266,7 @@ public class FishingHook extends Projectile {
         boolean flag = itemstack.is(Items.FISHING_ROD);
         boolean flag1 = itemstack1.is(Items.FISHING_ROD);
 
-        if (!player.isRemoved() && player.isAlive() && (flag || flag1) && this.distanceToSqr((Entity) player) <= 1024.0D) {
+        if (!player.isRemoved() && player.isAlive() && (flag || flag1) && this.distanceToSqr((Entity) player) <= me.titaniumtown.JettPackConfig.fishingRodLength) { // Jettpack
             return false;
         } else {
             this.discard();
