From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: satvrn <pastawho@gmail.com>
Date: Wed, 27 Sep 2023 22:08:34 -0700
Subject: [PATCH] Lithium: ai.raid


diff --git a/src/main/java/net/minecraft/world/entity/raid/Raid.java b/src/main/java/net/minecraft/world/entity/raid/Raid.java
index eaa2943b667967f93f28d9d794d702fdaeb670ec..c5df387a39ad342c20cb7890b4464655a387c7bf 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raid.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raid.java
@@ -62,6 +62,24 @@ import net.minecraft.world.phys.Vec3;
 
 public class Raid {
 
+    // Lithium start - ai.raid
+    private boolean isBarDirty;
+    private static final ItemStack CACHED_OMINOUS_BANNER;
+
+    static {
+        ItemStack itemstack = new ItemStack(Items.WHITE_BANNER);
+        CompoundTag nbttagcompound = new CompoundTag();
+        ListTag nbttaglist = (new BannerPattern.Builder()).addPattern(BannerPatterns.RHOMBUS_MIDDLE, DyeColor.CYAN).addPattern(BannerPatterns.STRIPE_BOTTOM, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.STRIPE_CENTER, DyeColor.GRAY).addPattern(BannerPatterns.BORDER, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.STRIPE_MIDDLE, DyeColor.BLACK).addPattern(BannerPatterns.HALF_HORIZONTAL, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.CIRCLE_MIDDLE, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.BORDER, DyeColor.BLACK).toListTag();
+
+        nbttagcompound.put("Patterns", nbttaglist);
+        BlockItem.setBlockEntityData(itemstack, BlockEntityType.BANNER, nbttagcompound);
+        itemstack.hideTooltipPart(ItemStack.TooltipPart.ADDITIONAL);
+        itemstack.setHoverName(Component.translatable("block.minecraft.ominous_banner").withStyle(ChatFormatting.GOLD));
+
+        CACHED_OMINOUS_BANNER = itemstack;
+    }
+    // Lithium end
+
     private static final int SECTION_RADIUS_FOR_FINDING_NEW_VILLAGE_CENTER = 2;
     private static final int ATTEMPT_RAID_FARTHEST = 0;
     private static final int ATTEMPT_RAID_CLOSE = 1;
@@ -272,6 +290,13 @@ public class Raid {
     }
 
     public void tick() {
+        // Lithium start - ai.raid
+        if (this.isBarDirty) {
+            this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
+            this.isBarDirty = false;
+        }
+        // Lithium end
+
         if (!this.isStopped()) {
             if (this.status == Raid.RaidStatus.ONGOING) {
                 boolean flag = this.active;
@@ -642,7 +667,7 @@ public class Raid {
     }
 
     public void updateBossbar() {
-        this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
+        this.isBarDirty = true; // Lithium - ai.raid
     }
 
     public float getHealthOfLivingRaiders() {
@@ -694,15 +719,7 @@ public class Raid {
     }
 
     public static ItemStack getLeaderBannerInstance() {
-        ItemStack itemstack = new ItemStack(Items.WHITE_BANNER);
-        CompoundTag nbttagcompound = new CompoundTag();
-        ListTag nbttaglist = (new BannerPattern.Builder()).addPattern(BannerPatterns.RHOMBUS_MIDDLE, DyeColor.CYAN).addPattern(BannerPatterns.STRIPE_BOTTOM, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.STRIPE_CENTER, DyeColor.GRAY).addPattern(BannerPatterns.BORDER, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.STRIPE_MIDDLE, DyeColor.BLACK).addPattern(BannerPatterns.HALF_HORIZONTAL, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.CIRCLE_MIDDLE, DyeColor.LIGHT_GRAY).addPattern(BannerPatterns.BORDER, DyeColor.BLACK).toListTag();
-
-        nbttagcompound.put("Patterns", nbttaglist);
-        BlockItem.setBlockEntityData(itemstack, BlockEntityType.BANNER, nbttagcompound);
-        itemstack.hideTooltipPart(ItemStack.TooltipPart.ADDITIONAL);
-        itemstack.setHoverName(Component.translatable("block.minecraft.ominous_banner").withStyle(ChatFormatting.GOLD));
-        return itemstack;
+        return CACHED_OMINOUS_BANNER; // Lithium - ai.raid
     }
 
     @Nullable
