From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: satvrn <pastawho@gmail.com>
Date: Mon, 25 Sep 2023 23:59:06 -0700
Subject: [PATCH] Carpet Fixes: Optimized getBiome call


diff --git a/src/main/java/net/minecraft/world/level/biome/BiomeManager.java b/src/main/java/net/minecraft/world/level/biome/BiomeManager.java
index 5695c5116c8a338b2e41aafcb2dc9f2146856970..51ecd31b709e526e0870519180672c042734c2c4 100644
--- a/src/main/java/net/minecraft/world/level/biome/BiomeManager.java
+++ b/src/main/java/net/minecraft/world/level/biome/BiomeManager.java
@@ -15,6 +15,8 @@ public class BiomeManager {
     private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
     private final long biomeZoomSeed;
 
+    private static final double maxOffset = 0.4500000001D; // Carpet Fixes
+
     public BiomeManager(BiomeManager.NoiseBiomeSource storage, long seed) {
         this.noiseBiomeSource = storage;
         this.biomeZoomSeed = seed;
@@ -45,12 +47,23 @@ public class BiomeManager {
             boolean bl = (p & 4) == 0;
             boolean bl2 = (p & 2) == 0;
             boolean bl3 = (p & 1) == 0;
-            int q = bl ? l : l + 1;
-            int r = bl2 ? m : m + 1;
-            int s = bl3 ? n : n + 1;
             double h = bl ? d : d - 1.0D;
             double t = bl2 ? e : e - 1.0D;
             double u = bl3 ? f : f - 1.0D;
+
+            // Carpet Fixes start - Optimized getBiome call
+            double maxQuartXX = 0.0D, maxQuartYY = 0.0D, maxQuartZZ = 0.0D;
+            if (p != 0) {
+                maxQuartXX = Mth.square(Math.max(h + maxOffset, Math.abs(h - maxOffset)));
+                maxQuartYY = Mth.square(Math.max(t + maxOffset, Math.abs(t - maxOffset)));
+                maxQuartZZ = Mth.square(Math.max(u + maxOffset, Math.abs(u - maxOffset)));
+                if (g < maxQuartXX + maxQuartYY + maxQuartZZ) continue;
+            }
+            // Carpet Fixes end
+
+            int q = bl ? l : l + 1;
+            int r = bl2 ? m : m + 1;
+            int s = bl3 ? n : n + 1;
             double v = getFiddledDistance(this.biomeZoomSeed, q, r, s, h, t, u);
             if (g > v) {
                 o = p;
