From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 19 Oct 2021 11:14:51 -0400
Subject: [PATCH] configurable logging of player login location


diff --git a/src/main/java/me/titaniumtown/JettPackConfig.java b/src/main/java/me/titaniumtown/JettPackConfig.java
index d81564ed381cff61b978adfa0ab661ac22415a07..2217bc4a378866eaf224731685b759a27268b655 100644
--- a/src/main/java/me/titaniumtown/JettPackConfig.java
+++ b/src/main/java/me/titaniumtown/JettPackConfig.java
@@ -114,4 +114,9 @@ public class JettPackConfig {
         config.addDefault(path, def);
         return config.getString(path, config.getString(path));
     }
+
+    public static boolean logPlayerLoginLoc = true;
+    private static void general() {
+        logPlayerLoginLoc = getBoolean("settings.log-player-login-location", logPlayerLoginLoc);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 40d55cade76d0e8f39f91a322bc8a821877efa2c..302e356c035ad0fac1854dae114e0c3edb6f2262 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -483,7 +483,13 @@ public abstract class PlayerList {
         }
         // Paper end
         // CraftBukkit - Moved from above, added world
-        PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+        // JettPack start -  configurable logging of player login location
+        if (me.titaniumtown.JettPackConfig.logPlayerLoginLoc) {
+            PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+        } else {
+            PlayerList.LOGGER.info("{}[{}] logged in with entity id {}", player.getName().getString(), s1, player.getId());
+        }
+        // JettPack end
     }
 
     public void updateEntireScoreboard(ServerScoreboard scoreboard, ServerPlayer player) {
