From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: satvrn <pastawho@gmail.com>
Date: Tue, 26 Sep 2023 00:14:22 -0700
Subject: [PATCH] Graphene (lithium): math.sine_lut


diff --git a/src/main/java/net/minecraft/util/Mth.java b/src/main/java/net/minecraft/util/Mth.java
index 3af5f3306c830b8983aaba5b4e80cc5c8973f441..0863ec9830d74558e07e50aa0ef4267ddff955bf 100644
--- a/src/main/java/net/minecraft/util/Mth.java
+++ b/src/main/java/net/minecraft/util/Mth.java
@@ -9,6 +9,7 @@ import net.minecraft.core.Vec3i;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.math.NumberUtils;
+import me.jellysquid.mods.lithium.util.math.CompactSineLUT; // Graphene - lithium: math.sine_lut
 
 public class Mth {
     private static final long UUID_VERSION = 61440L;
@@ -23,7 +24,7 @@ public class Mth {
     public static final float EPSILON = 1.0E-5F;
     public static final float SQRT_OF_TWO = sqrt(2.0F);
     private static final float SIN_SCALE = 10430.378F;
-    private static final float[] SIN = Util.make(new float[65536], (sineTable) -> {
+    public static final float[] SIN = Util.make(new float[65536], (sineTable) -> { // Graphene - lithium: math.sine_lut
         for(int i = 0; i < sineTable.length; ++i) {
             sineTable[i] = (float)Math.sin((double)i * Math.PI * 2.0D / 65536.0D);
         }
@@ -39,11 +40,11 @@ public class Mth {
     private static final double[] COS_TAB = new double[257];
 
     public static float sin(float value) {
-        return SIN[(int)(value * 10430.378F) & '\uffff'];
+        return CompactSineLUT.sin(value); // Graphene - lithium: math.sine_lut
     }
 
     public static float cos(float value) {
-        return SIN[(int)(value * 10430.378F + 16384.0F) & '\uffff'];
+        return CompactSineLUT.cos(value); // Graphene - lithium: math.sine_lut
     }
 
     public static float sqrt(float value) {
