From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etil <81570777+etil2jz@users.noreply.github.com>
Date: Fri, 12 Nov 2021 20:31:50 +0100
Subject: [PATCH] Improve EntityTransformEvent cancellation handling


diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index e95210c6d..f44e7bfeb 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -893,8 +893,6 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
             }
             // Paper end

-            if (org.spigotmc.SpigotConfig.logVillagerDeaths) Villager.LOGGER.info("Villager {} was struck by lightning {}.", this, lightning); // Paper - move log down, event can cancel
-
             entitywitch.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
             entitywitch.finalizeSpawn(world, world.getCurrentDifficultyAt(entitywitch.blockPosition()), MobSpawnType.CONVERSION, (SpawnGroupData) null, (CompoundTag) null);
             entitywitch.setNoAi(this.isNoAi());
@@ -906,8 +904,10 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
             entitywitch.setPersistenceRequired();
             // CraftBukkit start
             if (CraftEventFactory.callEntityTransformEvent(this, entitywitch, EntityTransformEvent.TransformReason.LIGHTNING).isCancelled()) {
+                super.thunderHit(world, lightning); // Mirai - didn't cancel the lighting, just the transformation
                 return;
             }
+            if (org.spigotmc.SpigotConfig.logVillagerDeaths) Villager.LOGGER.info("Villager {} was struck by lightning {}.", this, lightning); // Paper - move log down, event can cancel - // Mirai - move even further down because if transformation is cancelled, the entity won't die
             world.addFreshEntityWithPassengers(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
             // CraftBukkit end
             this.releaseAllPois();
