From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Fri, 3 Sep 2021 22:19:57 +0100
Subject: [PATCH] Chickens Avoid Foxes


diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index 855a816dd..4f2d5f7ba 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -16,6 +16,7 @@ import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
+import net.minecraft.world.entity.ai.goal.AvoidEntityGoal; // Crunchy
 import net.minecraft.world.entity.ai.goal.BreedGoal;
 import net.minecraft.world.entity.ai.goal.FloatGoal;
 import net.minecraft.world.entity.ai.goal.FollowParentGoal;
@@ -101,6 +102,12 @@ public class Chicken extends Animal {
             this.goalSelector.addGoal(1, new PanicGoal(this, 1.4D));
         }
         // Purpur end
+        // Crunchy start - Chickens avoid foxes
+        if (this.level.grapheneConfig.chickenAvoidFox) {
+            this.goalSelector.addGoal(1, new AvoidEntityGoal<>(this, Fox.class, this.level.grapheneConfig.chickenAvoidFoxDistance,
+                    this.level.grapheneConfig.chickenAvoidFoxSlow, this.level.grapheneConfig.chickenAvoidFoxFast));
+        }
+        // Crunchy end - Chickens avoid foxes
     }

     @Override
