From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrFishCakes <finlayoxby@outlook.com>
Date: Tue, 9 Nov 2021 22:40:20 +0000
Subject: [PATCH] Player Beheading


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index b6f9c99d580d985f2b84efaa655d9ae40073e72b..70e9bbc887343e4b509df1c3cb3c84a209401a06 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -869,7 +869,19 @@ public class ServerPlayer extends Player {
             return;
         }
         // Paper end
+        // Crunchy start - Player Beheading
+        if (level.grapheneConfig.beheadingEnabled) {
+            final double chance = level.grapheneConfig.beheadingChance;
 
+            if (chance > 0D && random.nextDouble() <= chance) {
+                if (level.grapheneConfig.isValidBeheadingItem(getBukkitEntity().getKiller())) {
+                    org.bukkit.inventory.ItemStack skullItem = new com.notsatvrn.graphene.SkullBuilder().setOfflinePlayer(getBukkitEntity()).build();
+
+                    level.getWorld().dropItemNaturally(new org.bukkit.Location(level.getWorld(), getX(), getY(), getZ()), skullItem); // Drop separate to loot
+                }
+            }
+        }
+        // Crunchy end - Player Beheading
         // SPIGOT-943 - only call if they have an inventory open
         if (this.containerMenu != this.inventoryMenu) {
             this.closeContainer(org.bukkit.event.inventory.InventoryCloseEvent.Reason.DEATH); // Paper
