From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NyCode <nico@nycode.de>
Date: Tue, 27 Jul 2021 14:09:05 +0200
Subject: [PATCH] Implement IP Hiding


diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index dda698b0aba3f50527870fa1e49e2c162f4edbfd..f92d190b638509f9404cd37c7075f2f14fa7d5ab 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -9,6 +9,7 @@ import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.security.PrivateKey;
 import java.util.Arrays;
+import java.util.Objects; // Nyper
 import java.util.Random;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -223,6 +224,11 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener
     }

     public String getUserName() {
+        // Nyper start
+        if (com.notsatvrn.graphene.GrapheneConfig.hidePlayerIps) {
+            return Objects.toString(this.gameProfile);
+        }
+        // Nyper end
         // Paper start
         String ip = com.destroystokyo.paper.PaperConfig.logPlayerIpAddresses ? String.valueOf(this.connection.getRemoteAddress()) : "<ip address withheld>";
         return this.gameProfile != null ? this.gameProfile + " (" + ip + ")" : String.valueOf(ip);
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 457cd49e4cc7087c180a53e38ff9278b2aa57924..8fcc290d6744b73bd373f4683a60b8efdb8fde1e 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -8,6 +8,7 @@ import com.mojang.authlib.GameProfile;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import de.nycode.nyper.NyperConfig; // Nyper
 import io.netty.buffer.Unpooled;
 import io.papermc.paper.adventure.PaperAdventure;
 import java.io.File;
@@ -497,9 +498,21 @@ public abstract class PlayerList {
         // CraftBukkit - Moved from above, added world
         // JettPack start -  configurable logging of player login location
         if (me.titaniumtown.JettPackConfig.logPlayerLoginLoc) {
-            PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+            // Nyper start
+            if (NyperConfig.hidePlayerIps) {
+                PlayerList.LOGGER.info("{} logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+            } else {
+                PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", player.getName().getString(), s1, player.getId(), worldserver1.serverLevelData.getLevelName(), player.getX(), player.getY(), player.getZ());
+            }
+            // Nyper end
         } else {
-            PlayerList.LOGGER.info("{}[{}] logged in with entity id {}", player.getName().getString(), s1, player.getId());
+            // Nyper start
+            if (NyperConfig.hidePlayerIps) {
+                PlayerList.LOGGER.info("{} logged in with entity id {}", player.getName().getString(), player.getId());
+            } else {
+                PlayerList.LOGGER.info("{}[{}] logged in with entity id {}", player.getName().getString(), s1, player.getId());
+            }
+            // Nyper end
         }
         // JettPack end
     }
