From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Satyrn <satyrn@disroot.org>
Date: Sun, 12 Dec 2021 19:20:00 +0000
Subject: [PATCH] Configuration Stuff

diff --git a/src/main/java/gg/airplane/compat/ServerConfigurations.java b/src/main/java/gg/airplane/compat/ServerConfigurations.java
index 246d75cf7..d910eaf90 100644
--- a/src/main/java/gg/airplane/compat/ServerConfigurations.java
+++ b/src/main/java/gg/airplane/compat/ServerConfigurations.java
@@ -25,10 +25,12 @@ public class ServerConfigurations {
       "spigot.yml",
       "paper.yml",
       "airplane.yml",
+      "purpur.yml",
       "graphene.yml",
       "pufferfish.yml",
       "jettpack.yml",
-      "crunchy.yml"
+      "crunchy.yml",
+      "nyper.yml"
     };

     public static Map<String, String> getCleanCopies() throws IOException {
