From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Satyrn <satyrn@disroot.org>
Date: Sun, 12 Dec 2021 19:20:00 +0000
Subject: [PATCH] Configuration Stuff

diff --git a/src/main/java/co/aikar/timings/TimingsExport.java b/src/main/java/co/aikar/timings/TimingsExport.java
index 5c405c8e9..a4bfbc8ca 100644
--- a/src/main/java/co/aikar/timings/TimingsExport.java
+++ b/src/main/java/co/aikar/timings/TimingsExport.java
@@ -234,8 +234,11 @@ public class TimingsExport extends Thread {
             pair("airplane", mapAsJSON(gg.airplane.AirplaneConfig.getConfigCopy(), null)), // Airplane
             pair("purpur", mapAsJSON(Bukkit.spigot().getPurpurConfig(), null)), // Pufferfish
             // Purpur end
+            pair("jettpack", mapAsJSON(Bukkit.spigot().getJettPackConfig(), null)), // Jettpack
+            pair("graphene", mapAsJSON(Bukkit.spigot().getGrapheneConfig(), null)), // Graphene
             pair("pufferfish", mapAsJSON(gg.pufferfish.pufferfish.PufferfishConfig.getConfigCopy(), null)), // Pufferfish
-            pair("crunchy", mapAsJSON(Bukkit.spigot().getCrunchyConfig(), null)) // Crunchy
+            pair("crunchy", mapAsJSON(Bukkit.spigot().getCrunchyConfig(), null)), // Crunchy
+            pair("nyper", mapAsJSON(Bukkit.spigot().getNyperConfig(), null)) // Nyper
         ));

         new TimingsExport(listeners, parent, history).start();
diff --git a/src/main/java/gg/airplane/compat/ServerConfigurations.java b/src/main/java/gg/airplane/compat/ServerConfigurations.java
index 246d75cf7..d910eaf90 100644
--- a/src/main/java/gg/airplane/compat/ServerConfigurations.java
+++ b/src/main/java/gg/airplane/compat/ServerConfigurations.java
@@ -25,10 +25,12 @@ public class ServerConfigurations {
       "spigot.yml",
       "paper.yml",
       "airplane.yml",
+      "purpur.yml",
       "graphene.yml",
       "pufferfish.yml",
       "jettpack.yml",
-      "crunchy.yml"
+      "crunchy.yml",
+      "nyper.yml"
     };

     public static Map<String, String> getCleanCopies() throws IOException {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 582dd6e47..b69d9592b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2623,6 +2623,20 @@ public final class CraftServer implements Server {
         }
         // Crunchy end

+        // Nyper start
+        @Override
+        public YamlConfiguration getNyperConfig() {
+            return de.nycode.nyper.NyperConfig.config;
+        }
+        // Nyper end
+
+        // Jettpack start
+        @Override
+        public YamlConfiguration getJettPackConfig() {
+            return me.titaniumtown.JettPackConfig.config;
+        }
+        // Jettpack end
+
         @Override
         public void restart() {
             org.spigotmc.RestartCommand.restart();
